jax/interpreters/pxla.py

```diff
@@ -61,144 +61,138 @@
+  to a logical device mesh. The length of the tuple is equal to the rank of the mesh.
+  Each mesh dimension can either get partitions of data varying along one of the
+  sharded dimensions, or the data can be replicated. See the docstring of
+  `MeshDimAssignment` for more information.
   """
+  sharding: Tuple[AvalDimSharding, ...]
+  mesh_mapping: Tuple[MeshDimAssignment, ...]
 
   def __init__(self,
-               shards_per_axis: Tuple[int, ...],
-               is_axis_materialized: Tuple[bool, ...],
-               replication_factors: List[Tuple[int, int]]):
-    assert len(shards_per_axis) == len(is_axis_materialized)
-    self.shards_per_axis = shards_per_axis
-    self.is_axis_materialized = is_axis_materialized
-    self.replication_factors = replication_factors
-
-  def __eq__(self, other):
-    return (self.shards_per_axis == other.shards_per_axis and
-            self.is_axis_materialized == other.is_axis_materialized and
-            self.replication_factors == other.replication_factors)
-
-  def __repr__(self):
-    return ("ShardingSpec(shards_per_axis=%s, is_axis_materialized=%s, "
-            "replication_factors=%s)" %
-            (self.shards_per_axis, self.is_axis_materialized,
-             self.replication_factors))
-
+               sharding: Iterable[AvalDimSharding],
+               mesh_mapping: Iterable[MeshDimAssignment]):
+    self.sharding = tuple(sharding)
+    self.mesh_mapping = tuple(mesh_mapping)
+
+  def indices(self, shape: Tuple[int, ...]) -> np.ndarray:
+    """Returns NumPy-style indices corresponding to a sharding spec.
+
+    Args:
+      shape: The shape of the logical array being sharded.
+
+    Returns:
+      An ndarray with a NumPy-style index for each element of the logical mesh.
+      Each element is an int, a slice object with step=1, or a tuple thereof, to
+      be treated as an index into the full logical array.
+    """
+    assert len(shape) == len(self.sharding)
+
+    axis_indices: List[Sequence[Index]] = []
+    chunk_indices_shape = []
+    for dim, sharding in enumerate(self.sharding):
+      axis_size = shape[dim]
+      if sharding is None:
+        axis_indices.append([slice(None)])
+        # NOTE: We don't append unsharded dimensions here, because they do not appear
+        #       in the mesh mapping.
+      elif isinstance(sharding, Unstacked):
+        assert axis_size == sharding.size, f'{axis_size} != {sharding.size}'
+        axis_indices.append(range(axis_size))
+        chunk_indices_shape.append(axis_size)
+      elif isinstance(sharding, Sharded):
+        shard_size, ragged = divmod(axis_size, sharding.shards)
+        assert not ragged
+        axis_indices.append([slice(i * shard_size, (i + 1) * shard_size)
+                             for i in range(sharding.shards)])
+        chunk_indices_shape.append(sharding.shards)
+      else:
+        assert_unreachable(sharding)
+
+    chunk_indices = np.empty([prod(chunk_indices_shape)], dtype=np.object)
```

Comment:
A comment explaining what `chunk_indices` is would be helpful... maybe something like, "chunk_indices is an ndarray representing the sharded axes of the logical array, with each dimension having size equal to the number of shards across the corresponding logical array dimension, and each element containing the multi-dimensional index that is used to extract the corresponding shard of the logical array."

Reply:
I guess I would call this `shard_indices` :P Either way though.